### PR TITLE
Remove test runner and the mock system API

### LINKED ISSUE #1372: Allow writing application unit tests that panic
# Motivation

Testing application error scenarios is very useful. However, the current Wasm unit tests only support checking returned errors. That means assertions and panics can not be currently tested inside unit tests.

# Proposed Solutions

## Using Rust conventions

Fork the `webassembly_test` crate (maybe into `linera-wasm-test`) and update the procedural macro to mark exported functions of tests that have a `#[should_panic]` attribute.

## Simpler Implementation

Add a new function exported from the `linera-wasm-test-runner` binary that a test can call to say it will panic. The test runner can then handle the runtime error as a test success.

### PR DESCRIPTION
**Pull request by jvff**
## Motivation

Unit tests for Linera applications were quite different from usual Rust tests. They required the usage of a custom test runner binary and the [`webassembly_test`](https://docs.rs/webassembly-test/latest/webassembly_test/) crate. The implementation of the tests also used a separate mock system API, which was already missing some APIs and requires maintenance of both a host side and a guest side.

## Proposal

Run the unit tests on the host architecture, which doesn't need a custom test runner nor the `webassembly_test` dependency. To do so, the mock system API was replaced with a test version of the `ContractRuntime` and `SystemRuntime` types instead. Selecting which type should be used (real or mock) is based on activating the `linera-sdk/test` feature flag, which is already configured on application examples and in the template to create a new application.

## Test Plan

The Wasm application unit tests have been ported to use the new way of writing unit tests. But the new mock runtime types should be tested, so issue #1975 was opened to track that.

## Release Plan

This changes the SDK and how applications write tests, so:
- Need to bump the major/minor version number in the next release of the crates.
- Need to update the developer manual.

## Links

- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

Closes #1372

### CODE REVIEW

File: examples/counter/src/contract.rs [hunk 1]
```diff
@@ -84,21 +82,19 @@ mod tests {
...
-    // fn message() {
-    // let initial_value = 72_u64;
-    // let mut counter = create_and_instantiate_counter(initial_value);
-
-    // counter
-    // .execute_message(())
-    // .now_or_never()
-    // .expect("Execution of counter operation should not await anything");
+    #[test]
+    #[should_panic(expected = "Counter application doesn't support any cross-chain messages")]
```
Comment (ma2bd) [inline examples/counter/src/contract.rs:86]: nice!

File: linera-sdk/src/contract/test_runtime.rs [hunk 2]
```diff
@@ -0,0 +1,652 @@
...
+    data_types::{Amount, BlockHeight, Resources, SendMessageRequest, Timestamp},
+    identifiers::{Account, ApplicationId, ChainId, ChannelName, Destination, MessageId, Owner},
+    ownership::{ChainOwnership, CloseChainError},
+};
+use serde::Serialize;
+
+use crate::Contract;
+
+/// A mock of the common runtime to interface with the host executing the contract.
+pub struct ContractRuntime<Application>
```
Comment (ma2bd) [inline linera-sdk/src/contract/test_runtime.rs:22]: Should this be called a `MockContractRuntime`?

Comment (ma2bd) [reply linera-sdk/src/contract/test_runtime.rs:22]: Could we have both `mod runtime` and `mod test_runtime` declared in `mod.rs` (with appropriate flags)? Then, either `pub use self::runtime::ContractRuntime;` or `pub use self::test_runtime::MockContractRuntime as ContractRuntime;` depending on the compilation.

**Review by ma2bd — APPROVED** [review]

File: linera-sdk/src/contract/test_runtime.rs (hunk 2 above)
Comment (jvff) [reply linera-sdk/src/contract/test_runtime.rs]: Yes, good idea! Done.

**Pull request merged**